### PR TITLE
SocketsHttpHandler: Add a switch to allow non-ascii headers

### PR DESCRIPTION
**Pull request by antonfirsov**
.NET Core's `SocketHttpHandler` does not allow sending non-ASCII characters in HTTP headers according to RFC, however in reality services accept/require other characters/encodings.

In dotnet/runtime#38711 we are introducing a generic solution for .NET 5. This is a 3.1-only workaround exposing two switches to relax header validation globally before the first usage of `HttpClient`, enabling advanced users to send Latin-1 encoded headers:
- AppContext switch: `System.Net.Http.SocketsHttpHandler.AllowLatin1Headers`
- Environment variable: `DOTNET_SYSTEM_NET_HTTP_SOCKETSHTTPHANDLER_ALLOWLATIN1HEADERS`

###  Customer Impact

Unblock migration to .NET Core 3.1 for customers who need to send Latin-1 headers.

### Testing

The PR adds tests (and minimal necessary test infrastructure changes) based on the ones in dotnet/runtime#39468 and dotnet/runtime#39169 to make sure these scenarios do not regress.

Validation: we sent private builds to the partner asking for this feature, and they confirmed that the change fixes their issue.

###  Risk

Low. The workaround is hidden behind a switch, our default behavior remains unchanged.

### CODE REVIEW

File: src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HPack/HPackEncoder.cs
Comment (Tratcher) [inline src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HPack/HPackEncoder.cs:223]: You should still filter by `0xFF00`. Downcasting a character in that range to a byte will cause data corruption. Only the 0xFF range can be cast to a byte.

**Review by Tratcher — DISMISSED** [review]

**Comment by antonfirsov** [comment]
~I realized my current tests are far from being correct. Converting to draft until I fix them & address the review comment from Chris.~

Done

File: src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs
Comment (antonfirsov) [inline src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs:1156]: I wanted to keep this change as minimal as possible, and did not change the exception message `Request headers must contain only ASCII characters.`, neither bothered to introduce a new one for the Latin1 case.

The message might be confusing when `AllowNonAsciiHeaders == true`, but only a minimal amount of (very advanced) users will be affected by this, therefore it's probably not worth the additional complexity and localization effort.

File: src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/StaticHttpSettings.cs
Comment (Tratcher) [inline src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/StaticHttpSettings.cs]: Would these all be more accurate as `AllowLatin1Headers`?

Comment (antonfirsov) [reply src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/StaticHttpSettings.cs]: In theory it could be any single-byte `ISO-8859-*`  encoding, not just Latin1, but probably still makes more sense, than "non-ascii".

Comment (antonfirsov) [reply src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/StaticHttpSettings.cs]: Renamed.

File: src/Common/tests/System/Net/Http/Http2LoopbackConnection.cs
Comment (ManickaP) [inline src/Common/tests/System/Net/Http/Http2LoopbackConnection.cs]: The same static variable is defined here: https://github.com/antonfirsov/corefx/blob/af/http-allow-nonascii/src/Common/tests/System/Net/Http/LoopbackServer.cs#L21
Could this be put at some share place?

File: src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/StaticHttpSettings.cs
Comment (ManickaP) [inline src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/StaticHttpSettings.cs:7]: I know you extracted this into a separate class because it was not building. We usually use `partial` class and `*.platform.cs` filename.

File: src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/StaticHttpSettings.cs
Comment (ManickaP) [inline src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/StaticHttpSettings.cs]: You renamed the constant but left the variable name with `NonAscii` is that on purpose?

File: src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Headers.cs
Comment (ManickaP) [inline src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Headers.cs]: NIT: indentation.

File: src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Headers.cs
Comment (ManickaP) [inline src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Headers.cs]: Isn't there a constant (`s_latin1Encoding`) somewhere introduce?

File: src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Headers.cs
Comment (ManickaP) [inline src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Headers.cs]:
```suggestion
                    catch (Exception) when (!expectSuccess)
                    { } // Eat up expected exceptions
```

File: src/System.Net.Http/tests/FunctionalTests/SocketsHttpHandlerTest.cs
Comment (ManickaP) [inline src/System.Net.Http/tests/FunctionalTests/SocketsHttpHandlerTest.cs:2264]: It can invoke an `async` method, we certainly do that in other places.

File: src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Headers.cs
Comment (ManickaP) [inline src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Headers.cs]: `RemoteExecutor` can invoke an async method, we certainly do that in other places.

File: src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Headers.cs
Comment (ManickaP) [inline src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Headers.cs:379]: It doesn't throw in this case?

File: src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Headers.cs
Comment (ManickaP) [inline src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Headers.cs:413]: It doesn't throw in this case?

File: src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HPack/HPackEncoder.cs
Comment (ManickaP) [inline src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HPack/HPackEncoder.cs]: NIT: extra empty line

File: src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs
Comment (ManickaP) [reply src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs:1156]: You could parametrize the message:
`Request headers must contain only {0} characters.` and pass ASCII or LATIN1

File: src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/StaticHttpSettings.cs
Comment (antonfirsov) [reply src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/StaticHttpSettings.cs:7]: Not only. It logically does not fit into `HttpConnectionSettings`, because that class contains contextual settings, while this setting is global.

File: src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/StaticHttpSettings.cs
Comment (antonfirsov) [reply src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/StaticHttpSettings.cs]: Good point, will fix!

File: src/Common/tests/System/Net/Http/Http2LoopbackConnection.cs
Comment (antonfirsov) [reply src/Common/tests/System/Net/Http/Http2LoopbackConnection.cs]: I considered `GenericLoopbackServer` for that, but felt like polluting a base type with unrelated implementation details. Any suggestion for a better place?

File: src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs
Comment (antonfirsov) [reply src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs:1156]: Aren't those messages localized? If yes, formatting may lead to issues unless translations are also updated.

File: src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Headers.cs
Comment (antonfirsov) [reply src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Headers.cs:379]: We swallow the exception in this case.

I can also imagine a case when `expectSuccess == false` but `await server.HandleRequestAsync()` does not throw, so returning here seemed the most robust solution here.

File: src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Headers.cs
Comment (antonfirsov) [reply src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Headers.cs:413]: Not in the receive test. It's actually only a regression test (will also succeed without my product code changes).

In this test, `expectSuccess` is only an indicator whether we should validate for Latin-1. Will think if I can make it easier to follow.

File: src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Headers.cs
Comment (antonfirsov) [reply src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Headers.cs]: I've no idea how and why did these lines change.

File: src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/StaticHttpSettings.cs
Comment (stephentoub) [inline src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/StaticHttpSettings.cs]: I don't think this needs to use Lazy.  Any code that accesses StaticHttpSettings is going to trigger this, so it may as well just be:
```C#
private static readonly bool s_allowLatin1Headers = GetAllowLatin1HeadersSetting();
```

File: src/Common/tests/System/Net/Http/Http2LoopbackConnection.cs
Comment (ManickaP) [reply src/Common/tests/System/Net/Http/Http2LoopbackConnection.cs]: You can put it in [`HttpHeaderData`](https://github.com/antonfirsov/corefx/blob/af/http-allow-nonascii/src/Common/tests/System/Net/Http/GenericLoopbackServer.cs#L84).

File: src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/StaticHttpSettings.cs
Comment (antonfirsov) [reply src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/StaticHttpSettings.cs]: I was afraid that type loading may trigger even earlier, preventing in-process calls of `AppContext.SetSwitch()` and `Environment.SetEnvironmentVariable()` to take effect.

**Review by MihaZupan — APPROVED** [review]
Change LGTM

File: src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/StaticHttpSettings.cs
Comment (antonfirsov) [reply src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/StaticHttpSettings.cs]: The tests seem to work fine after removing Lazy, so I pushed the change.

File: src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs
Comment (karelz) [inline src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs]: Should we change `c` to `int` as in other cases?

**Comment by antonfirsov** [comment]
The [Linux failure](https://dev.azure.com/dnceng/public/_build/results?buildId=807235&view=ms.vss-test-web.build-test-results-tab&runId=25605780&resultId=175618&paneView=debug) in System.Text.RegularExpressions.Tests is unrelated.

**Comment by antonfirsov** [comment]
Validation: We sent private binaries to our customer, and they confirmed that the change fixes their problem.

**Review by karelz — CHANGES_REQUESTED** [review]

File: src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs
Comment (antonfirsov) [reply src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs]: I did not like the implicit casting happening in the `foreach` expression. If anything. I would prefer rather `((int)c & mask) != 0` to make it more explicit.

Comment (antonfirsov) [reply src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs]: But this is no strong opinion, may change to `foreach (int c in s)` as well.

Comment (antonfirsov) [reply src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs]: In the end decided to change for consistency 😄

**Review by ManickaP — APPROVED** [review]
LGTM, thanks!

**Review by karelz — APPROVED** [review]

**Comment by antonfirsov** [comment]
No test failures are related to the change. (`System.Text.RegularExpressions.Tests` and `System.Diagnostics.Tests.PerformanceDataTests` do not use System.Net.Http)

**Comment by antonfirsov** [comment]
@danmosemsft can you help pushing this forward for servicing approval?

**Comment by danmoseley** [comment]
@antonfirsov usual procedure - create a port PR, @karelz supports it, let me know..

**Comment by danmoseley** [comment]
Oops, this is the PR. Got used to seeing 3.1 in the title.

File: src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HPack/HPackEncoder.cs
Comment (danmoseley) [reply src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HPack/HPackEncoder.cs:223]: Is this addressed?

File: src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Headers.cs
Comment (danmoseley) [inline src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Headers.cs:322]: Is this running in a child process? We generally don't want to set env vars in the test process as it affects other concurrent tests.

File: src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HPack/HPackEncoder.cs
Comment (Tratcher) [reply src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HPack/HPackEncoder.cs:223]: It was at some point but it seems to have regressed again.

Comment (Tratcher) [reply src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HPack/HPackEncoder.cs:223]: Nevermind, this links to old code.

Comment (danmoseley) [reply src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HPack/HPackEncoder.cs:223]: Ah looks like yes.

File: src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Headers.cs
Comment (Tratcher) [reply src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Headers.cs:322]: Or AppContext, that's effectively static too.

File: src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs
Comment (danmoseley) [reply src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs:1156]: To answer the question - no we do not localize any resources in netcoreapp. It's fine to change them at any time. If we didn't want to create new messages, we would probably normally extend the existing one so at least it wasn't definitely wrong. But not suggesting you change it now.

File: src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Headers.cs
Comment (danmoseley) [reply src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Headers.cs:322]: Right - for environment variables at least we would use RemoteInvoke. Otherwise I would expect potentially unreliable test results.

**Review by danmoseley — CHANGES_REQUESTED** [review]
Please double check that tests that set appcontext/set environment variables aren't affecting other tests in the same process.

**Comment by antonfirsov** [comment]
All tests are isolated using `RemoteExecutor`. The test methods are invoked by derived classes. Couldn't define them in the base class because of `RemoteExecutor` limitations -- it's not able to invoke delegates defined in an abstract class, since it has no info about what to instantiate.

**Review by danmoseley — APPROVED** [review]
I see it - thanks

**Comment by karelz** [comment]
@danmosemsft now that it is approved -- can we merge it, or should we wait for someone to tell us to do that?

**Comment by danmoseley** [comment]
You can go ahead and merge 5.0 PR's when marked servicing-approved plus the usual requirements - code reviewed and tests are good.

File: src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs
Comment (ManickaP) [reply src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs:1156]: In the light of this information, are you still against changing the exception message? @antonfirsov

**Comment by antonfirsov** [comment]
> You can go ahead and merge 5.0 PR's 

I hope this is also true for 3.1-only servicing PR-s :)

File: src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs
Comment (ManickaP) [inline src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs:1196]: NIT: Are all the new lines intentional here?

**Review by ManickaP — APPROVED** [review]
Little nits, otherwise LGTM, thanks!

**Comment by danmoseley** [comment]
> I hope this is also true for 3.1-only servicing PR-s :)

Oops, my bad. For 3.1/2.1, we normally have @Anipik do the merge, as he knows when the branch is open. Let him know if this is ready to merge.

File: src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs
Comment (antonfirsov) [reply src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs:1196]: Yes, I like to format code like this, but unsure what is the preferred BCL style.

File: src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs
Comment (antonfirsov) [reply src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs:1156]: Honestly, not convinced that it's worth the efforts to change this in the 3.1 branch (->rerunning all the tests, making sure nothing broken, etc). We expect very few customers to utilize this feature. (Most likely only 1)

**Review by Anipik — APPROVED** [review]
No Packaging changes required here. We can merge this once the feedback is addressed the ci is green

**Comment by danmoseley** [comment]
Ah yes that's the other reason it's good for @Anipik to merge the 2.1/3.1 changes -- because he helps check the packaging was done correctly - something we've gotten wrong more than once.

File: src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Headers.cs
Comment (antonfirsov) [reply src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Headers.cs:322]: Note: this got answered in a later https://github.com/dotnet/corefx/pull/42978#issuecomment-691257429

File: src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs
Comment (antonfirsov) [reply src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs:1156]: Note: we decided offline to not change this.

File: src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs
Comment (antonfirsov) [reply src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs:1196]: Note: we decided offline to not change this.

**Comment by antonfirsov** [comment]
@Anipik this PR is ready to merge:
important feedback was addressed, changes got approval, failures are unrelated flaky tests (System.Text.RegularExpressions.Tests and System.Diagnostics.Tests.PerformanceDataTests do not use System.Net.Http).

**Pull request merged**
